### PR TITLE
rework Desktop authorization layers

### PR DESCRIPTION
**Pull request by kabentley**
For desktop and mobile applications, all authentication happens on the backend. The frontend process merely initiates the login process and waits for notification that it succeeds. Previously the steps required to set up the process were somewhat complicated.

Now, to configure your electron or mobile application for authorization, pass the `authConfig` option to `ElectronApp.startup` or `IOSApp.startup` to specify your authorization configuration.

### CODE REVIEW

**Review by ramanujam-raman — APPROVED** [review]

**Review by paulius-valiunas — APPROVED** [review]

File: common/api/summary/ecschema-metadata.exports.csv [hunk 1]
```diff
@@ -50,7 +50,7 @@ beta;diagnosticCategoryToString(category: DiagnosticCategory): "Error" | "Warnin
 beta;DiagnosticCodes:
 beta;Diagnostics:
 beta;DiagnosticType
-beta;diagnosticTypeToString(type: DiagnosticType): "Schema" | "None" | "CustomAttributeContainer" | "Property" | "RelationshipConstraint" | "SchemaItem"
```
Comment (calebmshafer) [inline common/api/summary/ecschema-metadata.exports.csv:53]: Worried this may fail on the build machine...

**Comment by pmconne** [comment]
Do your changes warrant information in NextVersion.md? Is there example code somewhere?

File: core/common/src/BriefcaseTypes.ts [hunk 2]
```diff
@@ -25,10 +25,16 @@ export enum DownloadBriefcaseStatus {
...
+/** Whether a briefcase is editable or may only accept incoming changesets from iModelHub
  * @public
  */
-export enum SyncMode { FixedVersion = 1, PullAndPush = 2, PullOnly = 3 }
+export enum SyncMode {
+  FixedVersion = 1,
+  /** create a briefcase that can be edited. A unique briefcaseId is assigned by iModelHub */
+  PullAndPush = 2,
+  /** use [BriefcaseIdValue.Standalone](%backend). This makes a briefcase that can accept changesets from iModelHub but can never create changesets. */
+  PullOnly = 3
```
Comment (calebmshafer) [inline core/common/src/BriefcaseTypes.ts:36]: need trailing `,` for linter

**Review by khanaffan — APPROVED** [review]

File: core/electron-manager/src/backend/ElectronAuthorizationRequestHandler.ts [hunk 3]
```diff
@@ -15,10 +15,9 @@ import {
   AuthorizationResponseJson, AuthorizationServiceConfiguration, BasicQueryStringUtils,
 } from "@openid/appauth";
 import { NodeCrypto } from "@openid/appauth/built/node_support";
-import { BackendLoggerCategory } from "../BackendLoggerCategory";
 import { ElectronAuthorizationEvents } from "./ElectronAuthorizationEvents";
 
-const loggerCategory = BackendLoggerCategory.Authorization;
+const loggerCategory = "electron-backend";
```
Comment (calebmshafer) [inline core/electron-manager/src/backend/ElectronAuthorizationRequestHandler.ts:20]: Do you want to add this to the enum so it can be configured more easily?

**Review by calebmshafer — COMMENTED** [review]
cc: @kckst8 FYI a handful more changes that will need to be reflected in the iTwin Viewer update.

File: common/api/summary/ecschema-metadata.exports.csv (hunk 1 above)
Comment (kabentley) [reply common/api/summary/ecschema-metadata.exports.csv:53]: ugh, why does this keep getting into my changes? I didn't change anything in that package.

File: core/common/src/BriefcaseTypes.ts (hunk 2 above)
Comment (kabentley) [reply core/common/src/BriefcaseTypes.ts:36]: OK, but `npm run lint` doesn't flag this on my box. Do I have something wrong?

File: core/mobile-manager/src/backend/MobileHost.ts [hunk 4]
```diff
@@ -134,8 +99,18 @@ export class MobileHost {
     return new Promise<void>((resolve, reject) => {
 
       let progressCb: MobileProgressCallback | undefined;
+      let lastReportedOn = Date.now();
+      const minTimeBeforeReportingProgress = 1000;
       if (progress) {
         progressCb = (_bytesWritten: number, totalBytesWritten: number, totalBytesExpectedToWrite: number) => {
+          const currentTime = Date.now();
```
Comment (kabentley) [inline core/mobile-manager/src/backend/MobileHost.ts:106]: @khanaffan I think this logic should be moved to `NativeHost.downloadBriefcase`

File: core/common/src/BriefcaseTypes.ts (hunk 2 above)
Comment (calebmshafer) [reply core/common/src/BriefcaseTypes.ts:36]: Strange, maybe we changed that lint rule to not require this anymore...

**Review by DanEastBentley — APPROVED** [review]

File: core/mobile-manager/src/backend/MobileHost.ts (hunk 4 above)
Comment (kabentley) [reply core/mobile-manager/src/backend/MobileHost.ts:106]: but i don't understand, when is this used? For briefcases? How?

Comment (khanaffan) [reply core/mobile-manager/src/backend/MobileHost.ts:106]: This is called by MobileFileHandler. Which is set as file handler in iModelClient which is what is used to download changeset, checkpoints and briefcases.

I can move this logic down a step in native iOS level as it is the one firing up event more quickly then it could be transported to frontend.

Just one question is it possible for IPC notification. Progress kind of events are such that only latest event matters. If previous events are not process by frontend end it is fine to override it with latest. I did implement something similar when event was polled and progress was fired with flag to override previous event so poll will only find one event in queue at anytime for a given type of event.

Comment (pmconne) [reply core/mobile-manager/src/backend/MobileHost.ts:106]: Events are no longer queued for later dispatch, they are dispatched immediately. The frontend no longer polls for events. EventSink and EventSource no longer exist.

Comment (khanaffan) [reply core/mobile-manager/src/backend/MobileHost.ts:106]: So i guess this PR can go through as is, i will move this logic in separate PR down to IOS level to reduce calling into node in the first place. Just limit 1 progress event/ sec.

Comment (kabentley) [reply core/mobile-manager/src/backend/MobileHost.ts:106]: @khanaffan i added logic into the NativeHost.downloadBriefcase to only send progress events every 500 milliseconds. That takes care of the download briefcase use, is `MobileFileHandler` used for other types of downloads?

File: core/bentley/src/ProcessDetector.ts [hunk 5]
```diff
@@ -60,13 +60,13 @@ export class ProcessDetector {
    * @note this indicates that this is a browser process started by an iTwin mobile application.
    * It will return `false` when running user-launched web browsers on a mobile device.
    */
-  public static get isMobileAppFrontend() { return this.isBrowserProcess && window.location.origin === "imodeljs://app"; }
+  public static get isMobileAppFrontend() { return this.isAndroidAppFrontend || this.isIOSAppFrontend; }
```
Comment (kabentley) [inline core/bentley/src/ProcessDetector.ts:63]: @khanaffan why did you make these changes? I don't think they're right.

File: test-apps/ui-test-app/ios/imodeljs-test-app/imodeljs-test-app.xcodeproj/xcshareddata/xcschemes/imodeljs-test-app.xcscheme [hunk 6]
```diff
@@ -0,0 +1,85 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kabentley) [inline test-apps/ui-test-app/ios/imodeljs-test-app/imodeljs-test-app.xcodeproj/xcshareddata/xcschemes/imodeljs-test-app.xcscheme:1]: @khanaffan What are these changes?

File: test-apps/ui-test-app/ios/imodeljs-test-app/imodeljs-test-app/MainViewController.swift [hunk 7]
```diff
@@ -6,18 +6,71 @@
 import UIKit
 import WebKit
 
+class ProxyAuthClient : NSObject, AuthorizationClient {
```
Comment (kabentley) [inline test-apps/ui-test-app/ios/imodeljs-test-app/imodeljs-test-app/MainViewController.swift:9]: @khanaffan and here

File: core/mobile-manager/src/backend/MobileHost.ts (hunk 4 above)
Comment (khanaffan) [reply core/mobile-manager/src/backend/MobileHost.ts:106]: Yes it is. It also used for changeset download. But in that case we do not report progress to frontend. We have backlog item to report all progress for each of following step so app can actually show why its taking time to open a  iModel.

1. download briefcase
2. download changeset
3. apply changesets.

File: core/bentley/src/ProcessDetector.ts (hunk 5 above)
Comment (khanaffan) [reply core/bentley/src/ProcessDetector.ts]: As I explained this check is wrong. For debugging the source location could be http://<ip> or the custom url can change per app as two app should not have same custom url as its use for OIDC login as redirect url. We have determine the best way to ever tell about platform is just to pass it to frontend as argument. iOS App passes #port=<port> and platform=ios as argument to fronted. This is most reliable way to tell we have mobile frontend.

File: core/electron-manager/src/backend/ElectronAuthorizationRequestHandler.ts (hunk 3 above)
Comment (kabentley) [reply core/electron-manager/src/backend/ElectronAuthorizationRequestHandler.ts]: no, this is internal-only code and logged only to local desktops.

File: test-apps/ui-test-app/ios/imodeljs-test-app/imodeljs-test-app.xcodeproj/xcshareddata/xcschemes/imodeljs-test-app.xcscheme (hunk 6 above)
Comment (kabentley) [reply test-apps/ui-test-app/ios/imodeljs-test-app/imodeljs-test-app.xcodeproj/xcshareddata/xcschemes/imodeljs-test-app.xcscheme:1]: @khanaffan can you please make sure these changes are intended?

File: common/api/imodeljs-backend.api.md [hunk 8]
```diff
@@ -3014,24 +3009,33 @@ export class IpcHost {
...
 }
 
 // @beta
-export interface IpcHostOptions {
-    socket?: IpcSocketBackend;
+export interface IpcHostOpts {
+    // (undocumented)
+    iModelHost?: IModelHostConfiguration;
+    // (undocumented)
+    ipcHost?: {
```
Comment (pmconne) [inline common/api/imodeljs-backend.api.md:3033]: We talked about how it's weird to be able to start up an IPC host/app without supplying a socket, right? Will that be addressed in a separate PR?

**Comment by pmconne** [comment]
There are pre-existing tests that exercise this stuff, right?

File: core/backend/src/IpcHost.ts [hunk 9]
```diff
@@ -8,30 +8,39 @@
...
+  * @beta
+  */
+export interface IpcHostOpts {
+  iModelHost?: IModelHostConfiguration;
+  ipcHost?: {
+    /** The Ipc socket to use for communications with frontend. Allows undefined only for headless tests. */
+    socket?: IpcSocketBackend;
+
+    /** don't send stack information on exceptions */
+    exceptions?: {
```
Comment (pmconne) [inline core/backend/src/IpcHost.ts:31]: If `exceptions` is undefined does `noStack` default to false?

File: common/api/imodeljs-backend.api.md (hunk 8 above)
Comment (kabentley) [reply common/api/imodeljs-backend.api.md:3033]: On the frontend, it does not make sense. See `IpcApp.statup` where the socket is a separate argument, not part of the options and cannot be undefined.

On the backend, as noted in the documentation, `IpcHost.startup` does allow an undefined socket for "headless" tests.

File: core/bentley/src/ProcessDetector.ts [hunk 10]
```diff
@@ -76,4 +76,7 @@ export class ProcessDetector {
 
   /**  Is this process a mobile app backend? */
   public static get isMobileAppBackend() { return this.isIOSAppBackend || this.isAndroidAppBackend; }
+
+  /** Is this process the frontend of an native (Electron or Mobile) app? */
```
Comment (pmconne) [inline core/bentley/src/ProcessDetector.ts:80]: Typo: "an native".

File: core/common/src/BriefcaseTypes.ts [hunk 11]
```diff
@@ -25,10 +25,16 @@ export enum DownloadBriefcaseStatus {
   Error,
 }
 
-/** Operations allowed when synchronizing changes between the Briefcase and iModelHub
+/** Whether a briefcase is editable or may only accept incoming changesets from iModelHub
  * @public
  */
-export enum SyncMode { FixedVersion = 1, PullAndPush = 2, PullOnly = 3 }
+export enum SyncMode {
+  FixedVersion = 1,
```
Comment (pmconne) [inline core/common/src/BriefcaseTypes.ts:32]: undocumented.

File: core/common/src/IpcAppProps.ts [hunk 12]
```diff
@@ -73,8 +83,10 @@ export interface IpcAppFunctions {
   /** Determine whether there are outstanding txns . */
   hasPendingTxns: (key: string) => Promise<boolean>;
 
-  pullAndMergeChanges: (key: string) => Promise<IModelConnectionProps>;
-  pushChanges: (key: string, description: string) => Promise<IModelConnectionProps>;
+  /** pull (and potentially merge local) changesets up to the specified version into this Briefcase. */
+  pullAndMergeChanges: (key: string, version?: IModelVersionProps) => Promise<void>;
+
+  pushChanges: (key: string, description: string) => Promise<string>;
```
Comment (pmconne) [inline core/common/src/IpcAppProps.ts:89]: What is the meaning of the return value?

**Comment by kabentley** [comment]
> There are pre-existing tests that exercise this stuff, right?

nowhere near enough. Although it is hard to test authentication since it requires integration testing. @ramanujam-raman is going to work on a separate PR adding more thorough testing.

File: core/common/src/IpcAppProps.ts (hunk 12 above)
Comment (pmconne) [reply core/common/src/IpcAppProps.ts:89]: Apparently it's the changesetId.

File: docs/changehistory/NextVersion.md [hunk 13]
```diff
@@ -47,36 +50,64 @@ with:
...
   public get channel() {
     const channel = IModelApp.tileAdmin.channels.get("my-unique-channel-name");
     assert(undefined !== channel);
     return channel;
   }
 ```
+
+### Authentication changes for Electron and Mobile apps
+
+For desktop and mobile applications, all authentication happens on the backend. The frontend process merely initiates the login process and waits for notification that it succeeds. Previously the steps required to set up the process was somewhat complicated.
```
Comment (pmconne) [inline docs/changehistory/NextVersion.md:95]: Inconsequential typo: "was" should be "were".

**Review by pmconne — COMMENTED** [review]
I will pull and test display-test-app.

File: core/common/src/IpcAppProps.ts (hunk 12 above)
Comment (kabentley) [reply core/common/src/IpcAppProps.ts:89]: see the "real" API at `BriefcaseConnection.pushChanges`. Really the @internal interface definition is not the place to document usage. It is misleading to have extensive documentation here since it's not intended for anyone to call directly (though sometimes the interface methods are only used internally, and then have some internal-only documentation).

File: core/common/src/BriefcaseTypes.ts (hunk 11 above)
Comment (kabentley) [reply core/common/src/BriefcaseTypes.ts:32]: i tried adding @deprecated here, but gave up. I'll add a comment to explain what it was and how to migrate to the new approach.

File: docs/changehistory/NextVersion.md (hunk 13 above)
Comment (kabentley) [reply docs/changehistory/NextVersion.md:95]: oops, i already fixed this when i copied this text into the PR description. I'll fix here.

**Review by pmconne — CHANGES_REQUESTED** [review]
This appears to have broken reality models in display-test-app.
set SVT_STANDALONE_SIGNIN=1
set SVT_STANDALONE_FILENAME=d:\bim\3mx\ClarkClassify.bim
Start electron version of display-test-app.

Sign-in succeeds, reality model displays.
Open a different iModel containing reality data, or reopen this same one.
Observe a stream of "Request ended with an error  400" and reality models fail to display.

Reload app => behavior above repeats (first iModel displays reality model correctly, subsequently no reality models display).
Not reproducible on master.

**Comment by calebmshafer** [comment]
> > There are pre-existing tests that exercise this stuff, right?
> 
> nowhere near enough. Although it is hard to test authentication since it requires integration testing. @ramanujam-raman is going to work on a separate PR adding more thorough testing.

@kabentley @ramanujam-raman FWIW @adamhock is starting to work on some core ElectronHost testing and then hopefully we can move into the integration testing phase.  @ramanujam-raman let us know if you start on this, perhaps we can help out.

**Comment by pmconne** [comment]
display-test-app does not work in the browser any more. On startup:
Unhandled Rejection (InvalidStateError): Failed to execute 'send' on 'WebSocket': Still in CONNECTING state.

**Comment by kabentley** [comment]
> This appears to have broken reality models in display-test-app.
> set SVT_STANDALONE_SIGNIN=1
> set SVT_STANDALONE_FILENAME=d:\bim\3mx\ClarkClassify.bim
> Start electron version of display-test-app.
> 
> Sign-in succeeds, reality model displays.
> Open a different iModel containing reality data, or reopen this same one.
> Observe a stream of "Request ended with an error 400" and reality models fail to display.
> 
> Reload app => behavior above repeats (first iModel displays reality model correctly, subsequently no reality models display).
> Not reproducible on master.

@ramanujam-raman can you please see if you can figure this out?

**Comment by kabentley** [comment]
> display-test-app does not work in the browser any more. On startup:
> Unhandled Rejection (InvalidStateError): Failed to execute 'send' on 'WebSocket': Still in CONNECTING state.
> Error recurs on reloading page.
> 
> ```
> Uncaught (in promise) DOMException: Failed to execute 'send' on 'WebSocket': Still in CONNECTING state.
>     at LocalTransport.send (http://localhost:3000/static/js/main.chunk.js:64028:22)
>     at IpcWebSocketFrontend.invoke (http://localhost:3000/static/js/main.chunk.js:38053:32)
>     at Function.invoke (http://localhost:3000/static/js/main.chunk.js:63868:25)
>     at Function.callIpcChannel (http://localhost:3000/static/js/main.chunk.js:63892:36)
>     at Function.callIpcHost (http://localhost:3000/static/js/main.chunk.js:63901:21)
>     at Function.create (http://localhost:3000/static/js/main.chunk.js:112827:81)
>     at Function.startup (http://localhost:3000/static/js/main.chunk.js:61897:54)
>     at Function.startup (http://localhost:3000/static/js/main.chunk.js:63906:37)
>     at Function.startup (http://localhost:3000/static/js/main.chunk.js:64041:31)
>     at Function.startup (http://localhost:3000/static/js/main.chunk.js:235388:88)
>     at dtaFrontendMain (http://localhost:3000/static/js/main.chunk.js:236171:61)
> ```

@swbsi can you please investigate?

File: core/bentley/src/ProcessDetector.ts (hunk 10 above)
Comment (kabentley) [reply core/bentley/src/ProcessDetector.ts]: fixed

File: core/common/src/BriefcaseTypes.ts (hunk 11 above)
Comment (kabentley) [reply core/common/src/BriefcaseTypes.ts:33]: added

File: docs/changehistory/NextVersion.md (hunk 13 above)
Comment (kabentley) [reply docs/changehistory/NextVersion.md]: fixed

File: core/common/src/IpcAppProps.ts (hunk 12 above)
Comment (kabentley) [reply core/common/src/IpcAppProps.ts:82]: added notes to see real api

**Comment by pmconne** [comment]
@swbsi fixed the browser version of display-test-app - thanks. However, @ramanujam-raman, if I set SVT_STANDALONE_SIGNIN=1, I observe an infinite spinner and no login prompt when starting up the app in the browser.

File: core/frontend/src/LocalhostIpcApp.ts [hunk 14]
```diff
@@ -12,20 +12,27 @@ import { WebViewerApp, WebViewerAppOpts } from "./WebViewerApp";
...
 
   public constructor(port: number) {
     super();
 
     this._client = new WebSocket(`ws://localhost:${port}/`);
 
+    this._client.addEventListener("open", () => {
+      const pending = this._pending!;
+      this._pending = undefined;
+      pending.forEach((m) => this.send(m));
```
Comment (kabentley) [inline core/frontend/src/LocalhostIpcApp.ts:25]: @swbsi should you drop the listener here?

Comment (swbsi) [reply core/frontend/src/LocalhostIpcApp.ts:25]: > @swbsi should you drop the listener here?

The socket instance can't be opened again, so I have never bothered to.

**Comment by pmconne** [comment]
Token expiry appears fixed. Browser still displaying infinite spinner on startup.

**Comment by pmconne** [comment]
1 Set SVT_STANDALONE_SIGNIN=1 and SVT_STANDALONE_FILENAME=d:\bim\3mx\ClarkClassify.bim
2 Start display-test-app in browser.
3 Be prompted to sign in.
4 Click on email address.
5 Receive alert reading "please sign in.".
6 Click OK
7 GOTO 5

**Comment by kabentley** [comment]
> @ramanujam-raman @kabentley claims you assert display-test-app is working in browser now with sign-in. Why is it not working for me?

Oops, my bad, was trying to make it re-attempt on failed login, but whiffed. But @ramanujam-raman, it works for me but always makes me click on my email address. Is that right? Why?

**Comment by pmconne** [comment]
@ramanujam-raman watched me reproduce this about a dozen times. Now all of a sudden it appears to work consistently. Let's blame IMS. More importantly, let's write some proper tests for this sign-in stuff.

**Review by pmconne — APPROVED** [review]
